src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs

```diff
@@ -478,12 +478,13 @@ protected internal override void UpdateSectionAlignment(int sectionIndex, int al
             foreach ((string name, SymbolDefinition definition) in definedSymbols)
             {
                 MachSection section = _sections[definition.SectionIndex];
+                ushort descriptor = (ushort) (section.SectionName == "hydrated" ? N_NO_DEAD_STRIP : 0);
```

Comment:
Without this change, dead stripping was crashing NativeAOT `SmokeTests/UnitTests`.
In the mentioned tests linker found an opportunity to remove the following symbols, reducing the size of the hydrated section, which in turn was destructive for the app to startup as the layout didn't match the dehydrated data.
```diff
*** backups/strip-old/bin/bin-syms.lst	Fri Jun  7 14:08:43 2024
--- backups/strip-new/bin/bin-syms.lst	Fri Jun  7 16:45:45 2024
***************
*** 15452,15457 ****
--- 15452,15458 ----
  __ZTV79System_Linq_Expressions_System_Linq_Expressions_Interpreter_NewArrayInstruction
  __ZTV79System_Linq_Expressions_System_Linq_Expressions_Interpreter_NotEqualInstruction
  __ZTV79System_Linq_System_Linq_Enumerable_ListWhereIterator_1<S_P_CoreLib_System_Type>
+ __ZTV79UnitTests_Generics_TestSimpleGVMScenarios_AnotherDerivedClass_1<System___Canon>
  __ZTV79UnitTests_Interfaces_TestVariantGenericDynamicStaticVirtualMethods_AbjectFail_1
  __ZTV79UnitTests_Interfaces_TestVariantGenericDynamicStaticVirtualMethods_FrobCaller_2
  __ZTV80Boxed_S_P_CoreLib_System_Diagnostics_CodeAnalysis_DynamicallyAccessedMemberTypes
***************
*** 15566,15571 ****
--- 15567,15573 ----
  __ZTV82System_Linq_Expressions_System_Linq_Expressions_Interpreter_OrInstruction_OrUInt32
  __ZTV82System_Linq_Expressions_System_Linq_Expressions_Interpreter_OrInstruction_OrUInt64
  __ZTV82System_Linq_Expressions_System_Linq_Expressions_Interpreter_TryCatchFinallyHandler
+ __ZTV82UnitTests_Generics_TestDefaultGenericVirtualInterfaceMethods_Bar_1<System___Canon>
  __ZTV82UnitTests_Generics_TestThreadStaticFieldAccess_TypeWithThreadStaticField_1<Object>
  __ZTV82UnitTests_Generics_TestThreadStaticFieldAccess_TypeWithThreadStaticField_1<String>
  __ZTV82UnitTests_Interfaces_TestGenericDynamicStaticVirtualMethods_IFoo_1<System___Canon>
***************
*** 15680,15685 ****
--- 15682,15688 ----
  __ZTV84UnitTests_Generics_TestConstrainedGvmCalls_Derived_2<System___Canon__System___Canon>
  __ZTV84UnitTests_Generics_TestGvmDependencyFromGenericLazy_GenericDerived_1<System___Canon>
  __ZTV84UnitTests_Generics_TestGvmLookups_GenericDerivedWithNonGenericBase_1<System___Canon>
+ __ZTV84UnitTests_Interfaces_TestStaticInterfaceMethods_SimpleGenericClass_1<System___Canon>
  __ZTV84__Array<System_Linq_Expressions_System_Linq_Expressions_Interpreter_GotoInstruction>
  __ZTV84__Array<System_Linq_Expressions_System_Linq_Expressions_Interpreter_LocalDefinition>
  __ZTV84__Array<UnitTests_Interfaces_TestVariantGenericDynamicStaticVirtualMethods_AtomBase>
***************
*** 15920,15925 ****
--- 15923,15929 ----
  __ZTV91Boxed_S_P_CoreLib_System_Span_1<System_Linq_Expressions_System_Linq_Expressions_SwitchCase>
  __ZTV91Boxed_S_P_TypeLoader_Internal_Runtime_TypeLoader_OptionalFieldsRuntimeBuilder_OptionalField
  __ZTV91Boxed_System_Linq_System_Collections_Generic_SegmentedArrayBuilder_1_Arrays<System___Canon>
+ __ZTV91Boxed_UnitTests_Interfaces_TestStaticInterfaceMethods_SimpleGenericStruct_1<System___Canon>
  __ZTV91S_P_CoreLib_System_Collections_Generic_ICollection_1<UnitTests_Interfaces_SpecialArrayBase>
  __ZTV91S_P_CoreLib_System_Collections_Generic_IEnumerable_1<UnitTests_Interfaces_SpecialArrayBase>
  __ZTV91S_P_CoreLib_System_Collections_Generic_IList_1<S_P_CoreLib_System_Globalization_CalendarId>
```

Reply:
Out of curiosity, what if we mark all symbols as `N_NO_DEAD_STRIP`? (perhaps even without `S_ATTR_NO_DEAD_STRIP`)

Reply:
> In the mentioned tests linker found an opportunity to remove the following symbols

That's very likely a compiler bug. I've put this in my personal backlog to look at because that's a size saving.

Reply:
> Out of curiosity, what if we mark all symbols as `N_NO_DEAD_STRIP`? (perhaps even without `S_ATTR_NO_DEAD_STRIP`)

Seems like marking all the symbols with `N_NO_DEAD_STRIP` and removing the `S_ATTR_NO_DEAD_STRIP` from sections gives the same behaviour. If you agree I can apply this change, as it wouldn't have any special casing (like for `hydrated` section).

Reply:
I like the `N_NO_DEAD_STRIP` approach.